packages/angular_devkit/core/src/virtual-fs/host/memory.ts

```diff
@@ -0,0 +1,200 @@
+  }
+
+  write(path: Path, content: FileBuffer): Observable<void> {
+    if (this._isDir(path)) {
+      return _throw(new PathIsDirectoryException(path));
+    }
+
+    const existed = this._cache.has(path);
+    this._cache.set(path, content);
+    this._updateWatchers(path, existed ? HostWatchEventType.Changed : HostWatchEventType.Created);
+
+    return Observable.empty<void>();
+  }
+  read(path: Path): Observable<FileBuffer> {
+    if (this._isDir(path)) {
+      return _throw(new PathIsDirectoryException(path));
+    }
+    const maybeBuffer = this._cache.get(path);
+    if (!maybeBuffer) {
+      return _throw(new FileDoesNotExistException(path));
+    } else {
+      return ArrayObservable.of(maybeBuffer);
+    }
+  }
+  delete(path: Path): Observable<void> {
+    if (this._isDir(path)) {
+      for (const [cachePath, _] of this._cache.entries()) {
+        if (path.startsWith(cachePath + NormalizedSep)) {
+          this._cache.delete(cachePath);
+        }
+      }
+    } else {
+      this._cache.delete(path);
+    }
+    this._updateWatchers(path, HostWatchEventType.Deleted);
+
+    return Observable.empty();
+  }
+  rename(from: Path, to: Path): Observable<void> {
+    if (!this._cache.has(from)) {
+      return _throw(new FileDoesNotExistException(from));
+    } else if (this._cache.has(to)) {
+      return _throw(new FileAlreadyExistException(from));
+    }
+    if (this._isDir(from)) {
+      for (const path of this._cache.keys()) {
+        if (path.startsWith(from + NormalizedSep)) {
+          const content = this._cache.get(path);
+          if (content) {
+            this._cache.set(join(to, NormalizedSep, path.slice(from.length)), content);
+          }
+        }
+      }
+    } else {
+      const content = this._cache.get(from);
+      if (content) {
+        this._cache.delete(from);
+        this._cache.set(to, content);
+      }
+    }
+
+    this._updateWatchers(from, HostWatchEventType.Renamed);
+
+    return Observable.empty();
+  }
+
+  list(path: Path): Observable<PathFragment[]> {
+    if (this._cache.has(path)) {
+      return _throw(new PathIsFileException(path));
+    }
+    const fragments = split(path);
+    const result = new Set<PathFragment>();
+    for (const p of this._cache.keys()) {
+      if (p.startsWith(path + NormalizedSep)) {
+        result.add(split(p)[fragments.length]);
+      }
+    }
+
+    return ArrayObservable.of([...result]);
+  }
+
+  exists(path: Path): Observable<boolean> {
+    return ArrayObservable.of(this._cache.has(path) || this._isDir(path));
+  }
+  isDirectory(path: Path): Observable<boolean> {
+    return ArrayObservable.of(this._isDir(path));
+  }
+  isFile(path: Path): Observable<boolean> {
+    return ArrayObservable.of(this._cache.has(path));
+  }
+
+  stats(_path: Path): Observable<Stats<{}>> | null {
+    return null;
+  }
+
+  watch(path: Path, options?: HostWatchOptions): Observable<HostWatchEvent> | null {
+    const subject = new Subject<HostWatchEvent>();
```

Comment:
This way each new watcher gets a new subject. Maybe this is a stupid question, but isn't there a way to have a single subject with new watchers subscribing to it and filtering events by according to options?

Reply:
We would have to get a new subject by matching watchers. I went down that path and it was too complicated. We should re-evaluate if it's a problem later.